src/main/kotlin/org/rust/ide/inspections/lints/RsPathUsageAnalysis.kt

```diff
@@ -0,0 +1,111 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.lints
+
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.CachedValue
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.infer.ResolvedPath
+import org.rust.lang.core.types.inference
+import org.rust.openapiext.processElementsWithMacros
+
+data class PathUsageMap(
+    val pathUsages: Map<String, Set<RsElement>>,
+    val traitUsages: Set<RsTraitItem>
+)
+
+private val PATH_USAGE_KEY: Key<CachedValue<PathUsageMap>> = Key.create("PATH_USAGE_KEY")
+
+val RsItemsOwner.pathUsage: PathUsageMap
+    get() = CachedValuesManager.getCachedValue(this, PATH_USAGE_KEY) {
+        val usages = calculatePathUsages(this)
+        CachedValueProvider.Result.create(usages, PsiModificationTracker.MODIFICATION_COUNT)
+    }
+
+private fun calculatePathUsages(owner: RsItemsOwner): PathUsageMap {
+    val directUsages = mutableMapOf<String, MutableSet<RsElement>>()
+    val traitUsages = mutableSetOf<RsTraitItem>()
+
+    for (child in owner.children) {
+        processElementsWithMacros(child) { element ->
+            handleElement(element, directUsages, traitUsages)
+        }
+    }
+
+    return PathUsageMap(directUsages, traitUsages)
+}
+
+private fun handleElement(
+    element: PsiElement,
+    directUsages: MutableMap<String, MutableSet<RsElement>>,
+    traitUsages: MutableSet<RsTraitItem>
+): Boolean {
+    fun addItem(name: String, item: RsElement) {
+        directUsages.getOrPut(name) { mutableSetOf() }.add(item)
+    }
+
+    if (!element.isEnabledByCfg) return false
+
+    return when (element) {
+        is RsModItem -> false
+        is RsUseItem -> {
```

Comment:
It is handled separately, because otherwise this:
```rust
use foo::{S};
```
would be seen as an usage of `S`, although it is not usage, it is just import. But it should be probably handled in a different way, I'll think about it.

Reply:
We can use `RsPath#qualifier`, which is aware of imports